### PR TITLE
:safety_vest: Add CandidateSlot and Booking validation methods

### PR DESCRIPTION
**Pull request by CalculusEnjoyer**
Add method that checks if CandidateSlots overlap with existing CandidateSlots.
Add methods that check if Booking overlap with existing bookings.
Add methods that check if Booking in CandidateSlot and InterviewerSlot boundaries.
Write tests for those methods
Add exceptions related to those validations.

### CODE REVIEW

**Comment by CalculusEnjoyer** [comment]
|File|Coverage [77.06%]|:green_apple:|
|:-|:-:|:-:|
|[WeekService.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/c9ae785ba80cc8a20e0de7a00c4838255606167c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservices%2FWeekService.java)|100%|:green_apple:|
|[BookingOutOfSlotException.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/c9ae785ba80cc8a20e0de7a00c4838255606167c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fexceptions%2FBookingOutOfSlotException.java)|100%|:green_apple:|
|[OverlappingBookingException.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/c9ae785ba80cc8a20e0de7a00c4838255606167c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fexceptions%2FOverlappingBookingException.java)|100%|:green_apple:|
|[UtilValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/c9ae785ba80cc8a20e0de7a00c4838255606167c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FUtilValidator.java)|99.01%|:green_apple:|
|[BookingValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/c9ae785ba80cc8a20e0de7a00c4838255606167c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FBookingValidator.java)|93.14%|:green_apple:|
|[CandidateSlotValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/c9ae785ba80cc8a20e0de7a00c4838255606167c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FCandidateSlotValidator.java)|91.3%|:green_apple:|
|[InterviewerValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/c9ae785ba80cc8a20e0de7a00c4838255606167c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FInterviewerValidator.java)|33.33%|:green_apple:|

|Total Project Coverage|47.93%|:green_apple:|
|:-|:-:|:-:|

**Comment by CalculusEnjoyer** [comment]
|File|Coverage [83.74%]|:green_apple:|
|:-|:-:|:-:|
|[CandidateSlotValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/17da76d2554c905e82295feeff5c11eb7ee414e6/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FCandidateSlotValidator.java)|100%|:green_apple:|
|[WeekService.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/17da76d2554c905e82295feeff5c11eb7ee414e6/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservices%2FWeekService.java)|100%|:green_apple:|
|[OverlappingSlotException.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/17da76d2554c905e82295feeff5c11eb7ee414e6/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fexceptions%2FOverlappingSlotException.java)|100%|:green_apple:|
|[BookingOutOfSlotException.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/17da76d2554c905e82295feeff5c11eb7ee414e6/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fexceptions%2FBookingOutOfSlotException.java)|100%|:green_apple:|
|[OverlappingBookingException.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/17da76d2554c905e82295feeff5c11eb7ee414e6/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fexceptions%2FOverlappingBookingException.java)|100%|:green_apple:|
|[BookingValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/17da76d2554c905e82295feeff5c11eb7ee414e6/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FBookingValidator.java)|98.89%|:green_apple:|
|[UtilValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/17da76d2554c905e82295feeff5c11eb7ee414e6/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FUtilValidator.java)|88.89%|:green_apple:|
|[InterviewerValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/17da76d2554c905e82295feeff5c11eb7ee414e6/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FInterviewerValidator.java)|45.45%|:green_apple:|

|Total Project Coverage|48.45%|:green_apple:|
|:-|:-:|:-:|

**Comment by CalculusEnjoyer** [comment]
|File|Coverage [71.89%]|:green_apple:|
|:-|:-:|:-:|
|[WeekService.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/59c5f921a9e213808485f9c3249ae6cfb45ceef5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservices%2FWeekService.java)|100%|:green_apple:|
|[OverlappingSlotException.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/59c5f921a9e213808485f9c3249ae6cfb45ceef5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fexceptions%2FOverlappingSlotException.java)|100%|:green_apple:|
|[BookingOutOfSlotException.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/59c5f921a9e213808485f9c3249ae6cfb45ceef5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fexceptions%2FBookingOutOfSlotException.java)|100%|:green_apple:|
|[OverlappingBookingException.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/59c5f921a9e213808485f9c3249ae6cfb45ceef5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fexceptions%2FOverlappingBookingException.java)|100%|:green_apple:|
|[BookingValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/59c5f921a9e213808485f9c3249ae6cfb45ceef5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FBookingValidator.java)|98.89%|:green_apple:|
|[UtilValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/59c5f921a9e213808485f9c3249ae6cfb45ceef5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FUtilValidator.java)|88.89%|:green_apple:|
|[CandidateSlotValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/59c5f921a9e213808485f9c3249ae6cfb45ceef5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FCandidateSlotValidator.java)|68.67%|:green_apple:|
|[InterviewerValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/59c5f921a9e213808485f9c3249ae6cfb45ceef5/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FInterviewerValidator.java)|31.45%|:green_apple:|

|Total Project Coverage|47.22%|:green_apple:|
|:-|:-:|:-:|

File: src/main/java/com/intellias/intellistart/interviewplanning/services/WeekService.java [hunk 1]
```diff
@@ -38,7 +38,7 @@ public int getNextWeekNum() {
    *
    * @return formatted week number for a given date
    */
-  public int getWeekNumFrom(LocalDate date) {
+  public static int getWeekNumFrom(LocalDate date) {
```
Comment (VladOSIO2) [inline src/main/java/com/intellias/intellistart/interviewplanning/services/WeekService.java:41]: If we are making those methods static, so we don't need an instance of a service, maybe we should move this class to utility, possibly with renaming to WeekUtil? Not a problem for this PR, just a thought for future refactoring

File: src/main/java/com/intellias/intellistart/interviewplanning/util/validation/UtilValidator.java [hunk 2]
```diff
@@ -33,4 +34,38 @@ private static boolean isValidTimeBoundary(LocalTime time) {
...
         && time.getHour() >= 8 && time.getHour() <= 22;
   }
+
+  /**
+   * Checks if two intervals overlap.
+   */
+  public static boolean areIntervalsOverLapping(LocalDateTime from1, LocalDateTime to1,
+      LocalDateTime from2, LocalDateTime to2) {
+    return isTimeInInterval(from1, from2, to2)
+        || isTimeInInterval(to2, from2, to2)
```
Comment (VladOSIO2) [inline src/main/java/com/intellias/intellistart/interviewplanning/util/validation/UtilValidator.java:44]: Is a  `isTimeInInterval(to2, from2, to2)` is correct here? It checks if to2 in the interval of (from2, to2), which is always false. Maybe the first argument should be `to1`?

Comment (CalculusEnjoyer) [reply src/main/java/com/intellias/intellistart/interviewplanning/util/validation/UtilValidator.java:44]: Oh, yeah, that is my bad, going to change it

**Comment by CalculusEnjoyer** [comment]
|File|Coverage [70.56%]|:green_apple:|
|:-|:-:|:-:|
|[WeekService.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/50207217ff2c9c0d381d10c7fce2de216c4be32e/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservices%2FWeekService.java)|100%|:green_apple:|
|[OverlappingSlotException.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/50207217ff2c9c0d381d10c7fce2de216c4be32e/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fexceptions%2FOverlappingSlotException.java)|100%|:green_apple:|
|[BookingOutOfSlotException.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/50207217ff2c9c0d381d10c7fce2de216c4be32e/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fexceptions%2FBookingOutOfSlotException.java)|100%|:green_apple:|
|[OverlappingBookingException.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/50207217ff2c9c0d381d10c7fce2de216c4be32e/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fexceptions%2FOverlappingBookingException.java)|100%|:green_apple:|
|[BookingValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/50207217ff2c9c0d381d10c7fce2de216c4be32e/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FBookingValidator.java)|98.89%|:green_apple:|
|[UtilValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/50207217ff2c9c0d381d10c7fce2de216c4be32e/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FUtilValidator.java)|81.89%|:green_apple:|
|[CandidateSlotValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/50207217ff2c9c0d381d10c7fce2de216c4be32e/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FCandidateSlotValidator.java)|68.67%|:green_apple:|
|[InterviewerValidator.java](https://github.com/CalculusEnjoyer/intellistart-java-2022-vanquishers/blob/50207217ff2c9c0d381d10c7fce2de216c4be32e/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2Fvalidation%2FInterviewerValidator.java)|31.45%|:green_apple:|

|Total Project Coverage|47.06%|:green_apple:|
|:-|:-:|:-:|

**Review by VladOSIO2 — APPROVED** [review]
OK

**Pull request merged**